// ngx-localstorage/esm5/lib/ngx-localstorage.module.js
import { __decorate, __metadata, __param } from "tslib";
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { LocalStorageDirective } from './directives/ngx-localstorage.directive';
import { NGX_LOCAL_STORAGE_CONFIG } from './tokens/storage-config';
import { DefaultSerializer } from './classes/default-serializer';
import { NGX_LOCAL_STORAGE_SERIALIZER } from './tokens/storage-serializer';
/**
 * Provides the librarys module.
 */
var NgxLocalStorageModule = /** @class */ (function () {
    /**
     * Creates a new instance.
     */
    function NgxLocalStorageModule(parentModule) {
        if (parentModule) {
            throw new Error('NgxLocalStorageModule is already loaded. Import it in the AppModule only');
        }
    }
    NgxLocalStorageModule_1 = NgxLocalStorageModule;
    /**
     * Initializes the module for the root module with the given configuration.
     */
    NgxLocalStorageModule.forRoot = function (config) {
        return {
            ngModule: NgxLocalStorageModule_1,
            providers: [
                {
                    provide: NGX_LOCAL_STORAGE_CONFIG,
                    useValue: config
                }
            ]
        };
    };
    var NgxLocalStorageModule_1;
    NgxLocalStorageModule.ctorParameters = function () { return [
        { type: NgxLocalStorageModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    NgxLocalStorageModule = NgxLocalStorageModule_1 = __decorate([
        NgModule({
            imports: [],
            declarations: [
                LocalStorageDirective
            ],
            exports: [
                LocalStorageDirective
            ],
            providers: [
                {
                    provide: NGX_LOCAL_STORAGE_SERIALIZER,
                    useClass: DefaultSerializer
                }
            ]
        }),
        __param(0, Optional()), __param(0, SkipSelf()),
        __metadata("design:paramtypes", [NgxLocalStorageModule])
    ], NgxLocalStorageModule);
    return NgxLocalStorageModule;
}());
export { NgxLocalStorageModule };
