// ngx-localstorage/esm5/lib/utils.js
/**
 * Gets an objects property based on its path.
 * @param path Path to the property
 * @param object Object to access
 */
export var getProperty = function (path, object) {
    return path.reduce(function (obj, p) { return (!!obj) ? obj[p] : null; }, object);
};
/**
 * Sets an objects property based on its path.
 * @param path Path to the property
 * @param value Value to set
 * @param object Object whose value to set
 * @param falsyTransformer optional transformer handling falsy values
 */
export var setProperty = function (path, value, object, falsyTransformer) {
    var lastKeyIndex = path.length - 1;
    for (var i = 0; i < lastKeyIndex; ++i) {
        var key = path[i];
        if (!(key in object)) {
            object[key] = {};
        }
        object = object[key];
    }
    object[path[lastKeyIndex]] = (!value || (typeof value === 'string' && value === 'false'))
        && !!falsyTransformer ? falsyTransformer() : value;
};
/**
 * Constructs the storage key based on a prefix - if given - and the key itself
 */
export var constructKey = function (key, prefix, configuredPrefix) {
    var prefixToUse = prefix || configuredPrefix;
    if (prefixToUse) {
        return prefixToUse + "_" + key;
    }
    return key;
};
/**
 * The librarys default config.
 */
export var defaultConfig = {
    allowNull: true
};
/**
 * StorageSerializer Guard
 */
export var isSerializer = function (prefixOrSerializer) {
    return !!prefixOrSerializer && prefixOrSerializer.serialize !== undefined;
};
