// ngx-localstorage/esm2015/lib/ngx-localstorage.module.js
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { LocalStorageDirective } from './directives/ngx-localstorage.directive';
import { NGX_LOCAL_STORAGE_CONFIG } from './tokens/storage-config';
import { DefaultSerializer } from './classes/default-serializer';
import { NGX_LOCAL_STORAGE_SERIALIZER } from './tokens/storage-serializer';
/**
 * Provides the librarys module.
 */
export class NgxLocalStorageModule {
    /**
     * Creates a new instance.
     */
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('NgxLocalStorageModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * Initializes the module for the root module with the given configuration.
     */
    static forRoot(config) {
        return {
            ngModule: NgxLocalStorageModule,
            providers: [
                {
                    provide: NGX_LOCAL_STORAGE_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
NgxLocalStorageModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [
                    LocalStorageDirective
                ],
                exports: [
                    LocalStorageDirective
                ],
                providers: [
                    {
                        provide: NGX_LOCAL_STORAGE_SERIALIZER,
                        useClass: DefaultSerializer
                    }
                ]
            },] }
];
NgxLocalStorageModule.ctorParameters = () => [
    { type: NgxLocalStorageModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
